tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,517 @@
+    }
+}
+
+impl LeaderTpuCacheService {
+    /// Run the [`LeaderTpuCacheService`], returning receiver and the service.
+    pub async fn run(
+        rpc_client: Arc<impl ClusterInfoProvider + 'static>,
+        mut slot_receiver: SlotReceiver,
+        config: LeaderTpuCacheServiceConfig,
+        cancel: CancellationToken,
+    ) -> Result<(LeaderUpdateReceiver, Self), LeaderTpuCacheServiceError> {
+        let (mut leader_tpu_map, mut epoch_info, mut slot_leaders) = Self::initialize_state(
+            rpc_client.as_ref(),
+            slot_receiver.clone(),
+            config.max_consecutive_failures,
+        )
+        .await?;
+        let (current_slot, lookahead_leaders) = get_slot_and_lookahead(
+            slot_receiver.slot(),
+            &slot_leaders,
+            config.lookahead_leaders,
+        );
+        let leaders = Self::leader_sockets(
+            current_slot,
+            lookahead_leaders,
+            &slot_leaders,
+            &leader_tpu_map,
+        );
+
+        let (leaders_sender, leaders_receiver) =
+            watch::channel((leaders, config.lookahead_leaders != lookahead_leaders));
+
+        let cancel_clone = cancel.clone();
+        let main_loop = async move {
+            let mut num_consequent_failures: usize = 0;
+            let mut refresh_tpu_interval = interval(config.refresh_nodes_info_every);
+            loop {
+                tokio::select! {
+                    _ = refresh_tpu_interval.tick() => {
+                        try_update(
+                            "cluster TPU ports",
+                            &mut leader_tpu_map,
+                            || LeaderTpuMap::new(rpc_client.as_ref()),
+                            &mut num_consequent_failures,
+                            config.max_consecutive_failures,
+                        ).await?;
+                        debug!("Updated cluster TPU ports");
+                    }
+                    res = slot_receiver.changed() => {
+                        debug!("Changed slot receiver");
+                        if let Err(e) = res {
+                            warn!("Slot receiver channel closed: {e}");
+                            break;
+                        }
+
+                        let estimated_current_slot = slot_receiver.slot().first_slot();
+                        if estimated_current_slot > epoch_info.last_slot_in_epoch {
+                            try_update(
+                                "epoch info",
+                                &mut epoch_info,
+                                || EpochInfo::new(rpc_client.as_ref(), estimated_current_slot),
+                                &mut num_consequent_failures,
+                                config.max_consecutive_failures,
+                            ).await?;
+                        }
+                        if estimated_current_slot > slot_leaders.last_slot().saturating_sub(MAX_FANOUT_SLOTS) {
```

Comment:
As slots are represented with a `u64` and start at `0`, it is probably better to use a `saturating_add()` rather than a `saturating_sub()`:

```suggestion
                        if estimated_current_slot.saturating_add(MAX_FANOUT_SLOTS) > slot_leaders.last_slot() {
```